element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java

```diff
@@ -0,0 +1,180 @@
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * TCP server connection is used by CoapEndpoint when instantiated by the CoapServer. Per RFC the server can both
+ * send and receive messages, but cannot initiated new outgoing connections.
+ */
+public class TcpServerConnector implements Connector {
+
+    private final static Logger LOGGER = Logger.getLogger(TcpServerConnector.class.getName());
+
+    private final int numberOfThreads;
+    private final InetSocketAddress localAddress;
+    private final int connectionIdleTimeoutSeconds;
+    private final ConcurrentMap<SocketAddress, Channel> activeChannels = new ConcurrentHashMap<>();
+
+    private RawDataChannel rawDataChannel;
+    private EventLoopGroup bossGroup;
+    private EventLoopGroup workerGroup;
+
+    public TcpServerConnector(InetSocketAddress localAddress, int idleTimeout, int numberOfThreads) {
+        this.numberOfThreads = numberOfThreads;
+        this.connectionIdleTimeoutSeconds = idleTimeout;
+        this.localAddress = localAddress;
+    }
+
+    @Override
+    public void start() throws IOException {
+        if (rawDataChannel == null) {
+            throw new IllegalStateException("Cannot start without message handler.");
+        }
+        if (bossGroup != null) {
+            throw new IllegalStateException("Connector already started");
+        }
+        if (workerGroup != null) {
+            throw new IllegalStateException("Connector already started");
+        }
+
+        bossGroup = new NioEventLoopGroup(1);
+        workerGroup = new NioEventLoopGroup(numberOfThreads);
+
+        ServerBootstrap bootstrap = new ServerBootstrap();
+        bootstrap.group(bossGroup, workerGroup)
+                .channel(NioServerSocketChannel.class)
+                .option(ChannelOption.SO_BACKLOG, 100)
+                .option(ChannelOption.SO_KEEPALIVE, true)
+                .option(ChannelOption.AUTO_READ, true)
+                .childHandler(new ChannelRegistry());
+
+        // Start the server.
+        bootstrap.bind(localAddress).syncUninterruptibly();
+    }
+
+    @Override
+    public void stop() {
+        bossGroup.shutdownGracefully(0, 1, TimeUnit.SECONDS).syncUninterruptibly();
+        workerGroup.shutdownGracefully(0, 1, TimeUnit.SECONDS).syncUninterruptibly();
+
+        workerGroup = null;
+        bossGroup = null;
+    }
+
+    @Override
+    public void destroy() {
+        stop();
+    }
+
+    @Override
+    public void send(RawData msg) {
+        Channel channel = activeChannels.get(msg.getInetSocketAddress());
+        if (channel == null) {
+            // TODO: Is it worth allowing opening a new connection when in server mode?
```

Comment:
Is it a problem to allow for opening an outbound connection? ;-)
I have to admit though that the obvious use cases (currently) should work with this kind of separation into `TcpServerConnector` and `TcpClientConnector`. In the DTLS case, however, the `DTLSConnector` is able to initiate a DTLS handshake no matter whether it is on the client or server side. 
The separation also seems to limit the user code to the role chosen by configuring either sevrer or client connector, doesn't it? The `CoapServer` allows for registering multiple `Endpoint`s. The one to use for communicating with a peer is currently chosen based on whether the remote URI is a non-secure or secure URI, i.e. we currently only differentiate between secure and non-secure endpoints. In order to be able to use UDP and TCP endpoints in parallel, we would need to introduce a way to pick the endpoint based on the URI scheme in general, i.e. a request with a URI with the `coap+tcp` scheme should be routed to a TCP endpoint whereas a simple `coaps` request should be routed to a UDP/DTLS endpoint.